packages/react-native/native-js-interface.md

```diff
@@ -4,27 +4,143 @@ This is a working spec of this the interface that the native layer (`BugsnagReac
 
 ## Intro
 
-`BugsnagReactNative` is a class which will be instantiated once for the entire runtime and accessible via JS. It's responsible for managing the instantiation of the native client (either `bugsnag-android` or `bugsnag-cocoa`) and providing forwarding on data and invocations from JS to the native client.
+`BugsnagReactNative` is a class which will be instantiated once for the entire runtime and accessible via JS. It's responsible for managing the communicating with the native client (either `bugsnag-android` or `bugsnag-cocoa`) and providing data and invocations from JS to the native client.
 
 ## Accessing `BugsnagReactNative` from JavaScript
 
 Access is set up in the standard React Native way<sup>[1](https://facebook.github.io/react-native/docs/0.59/native-modules-ios),[2](https://facebook.github.io/react-native/docs/0.59/native-modules-android)</sup> such that `BugsnagReactNative` is available in JavaScript like so:
 
 ```js
 const { NativeModules } = require('react-native')
-const NativeClient = NativeModules.BugsnagReactNative
+const { BugsnagReactNative } = NativeModules
 ```
 
 ## API
 
-### `BugsnagReactNative.leaveBreadcrumb(breadcrumb): void`
+### Methods
+
+#### `BugsnagReactNative.getConfig(): configMap`
+
+Retrieves a data structure representing the configuration which the native client was initialised with. The data structure needn't be immutable since it will be passed over the bridge anyway, but mutations should not have any effect.
+
+An "empty" map will indicate to the JS layer that the native client is not configured, which is a problem.
+
+This method is required to be synchronous so it should be annotated as such. This is because during its synchronous configuration, the JS layer needs to obtain config from the native layer.
+
+###### Android
+
+```java
+@ReactMethod(isBlockingSynchronousMethod = true)
+public ReadableMap getConfig()
+```
+
+###### iOS
+```objc
+(NSDictionary *)getConfig;
+
+// annotation goes in BugsnagReactNative.m
+RCT_EXPORT_BLOCKING_SYNCHRONOUS_METHOD(getConfig);
+```
+
+#### `BugsnagReactNative.updateClientState(key, updates): void`
+
+Updates to mutable client state in JS (e.g. `metaData`, `user`) should be replicated and stored in the native layer using this method.
+
+###### Android
+```java
+void updateClientState(String key, ReadableMap update)
+```
+
+###### iOS
+```objc
+- (void)updateClientState:(NSString * key)
+                         :newValue (NSDictionary *)update;
+```
+
+The keys in the updates map are the names of the properties to be updated and the values are encoded objects containing the new value for that property. The list of allowed keys is:
+
+- `user`: a Map<String, String> with `id`, `name` and `email`
+- `metaData`: a Map<String, Map> of metadata tabs
+
+#### `BugsnagReactNative.deliver(payload): Promise<boolean>`
```

Comment:
There are two things on my mind here:

1. we want to avoid synchronous bridge calls if at all possible (they block the main thread)
2. in handling a fatal JS error, my understanding is that the process remains alive until there is no work left to do – so any async stuff in an `onerror` callback is allowed to complete.

So I think this ok – but we can and should revisit when implementing and testing.

Reply:
It feels reasonable to block the main thread when generating an error report - but maybe I'm coming at this from a mobile rather than JS perspective. For context there were some edge cases around termination in the previous implementation, specifically: https://github.com/bugsnag/bugsnag-react-native/pull/290

Reply:
1. consider a better name. suggestion `notify()`
2. figure out if native threads are ever useful for JS exceptions, and if so, do we need them at the exact point the error happened (this will determine whether we need to make `getPayloadInfo()` synchronous.

Reply:
> It feels reasonable to block the main thread when generating an error report - but maybe I'm coming at this from a mobile rather than JS perspective.

We should use the new interface in a way that we can get around this requirement entirely - my thoughts for iOS are to bypass the existing delivery mechanism and use `-[BugsnagConfiguration session]` directly to send a request. That way, we can delay resolving the promise until the request completes (successfully or no), and the JS layer can then invoke the default fatal handler.

Reply:
I'm concerned that bypassing the existing delivery mechanism will lead to edge-cases/different behaviour around how JS reports are cached, and that this will introduce additional complexity to our report (de)serialisation code.

Calling `bugsnag.notify()` with the JS payload coerced into an exception/error would give consistent behaviour for a variety of things (callbacks, payload validation, delivery caching, etc), and seems less risky. Maybe we could investigate exposing an internal `notify` method that takes a `Report` object as a parameter, rather than an `Exception`?